core/src/main/scala/com.softwaremill.kmq/redelivery/ConsumeMarkersActor.scala

```diff
@@ -2,172 +2,223 @@ package com.softwaremill.kmq.redelivery
+        _ <- assignedPartitions.update(_ - p)
+      } yield ()
+
+    def partitionAssigned(
+        producer: KafkaProducer[Array[Byte], Array[Byte]],
+        p: Partition,
+        endOffset: Offset,
+        assignedPartitions: Ref[IO, Map[Partition, AssignedPartition]],
+        dispatcher: Dispatcher[IO]
+    ): IO[Unit] = for {
+      (redeliverActor, shutdown) <- RedeliverActor.create(p, producer, config, helpers.clients).allocated
+      _ <- redeliverActor.tell(DoRedeliver)
+      _ <- assignedPartitions.update(
+        _.updated(
+          p,
+          AssignedPartition(new MarkersQueue(endOffset - 1), redeliverActor, shutdown, config, None, None, dispatcher)
+        )
+      )
+    } yield ()
+
+    def setupOffsetCommitting = for {
+      commitMarkerOffsetsActor <- CommitMarkerOffsetsActor
+        .create(
+          config.getMarkerTopic,
+          config.getMarkerConsumerOffsetGroupId,
+          helpers
+        )
+      _ <- Resource.eval(commitMarkerOffsetsActor.tell(DoCommit))
+    } yield commitMarkerOffsetsActor
+
+    for {
+      dispatcher <- Dispatcher.sequential[IO]
+      assignedPartitions <- Resource.eval(IO.ref(Map.empty[Partition, AssignedPartition]))
+      mailbox <- Resource.eval(Queue.unbounded[IO, ConsumeMarkersActorMessage])
+      markerConsumer <- helpers.createConsumer(
+        config.getMarkerConsumerGroupId,
+        classOf[MarkerKey.MarkerKeyDeserializer],
+        classOf[MarkerValue.MarkerValueDeserializer]
+      )
+      _ <- setupMarkerConsumer(markerConsumer, assignedPartitions, dispatcher)
+      commitMarkerOffsetsActor <- setupOffsetCommitting
+      _ <- stream(assignedPartitions, mailbox, commitMarkerOffsetsActor, markerConsumer).compile.drain.background
+      _ <- Resource.eval(logger.info("Consume markers actor setup complete"))
+    } yield new ConsumeMarkersActor(mailbox)
   }
 
-  override def receive: Receive = {
-    case DoConsume =>
-      try {
-        val markers = markerConsumer.poll(OneSecond).asScala
-        val now = System.currentTimeMillis()
-
-        markers.groupBy(_.partition()).foreach { case (partition, records) =>
-          assignedPartitions.get(partition) match {
-            case None =>
-              throw new IllegalStateException(s"Got marker for partition $partition: ${records.map(_.key())}, but partition is not assigned!")
-
-            case Some(ap) =>
-              ap.handleRecords(records, now)
-              ap.markersQueue.smallestMarkerOffset().foreach { offset =>
-                commitMarkerOffsetsActor ! CommitOffset(partition, offset)
-              }
+  private def stream(
+      assignedPartitions: Ref[IO, Map[Partition, AssignedPartition]],
+      mailbox: Queue[IO, ConsumeMarkersActorMessage],
+      commitMarkerOffsetsActor: CommitMarkerOffsetsActor,
+      markerConsumer: KafkaConsumer[MarkerKey, MarkerValue]
+  ) = Stream
+    .fromQueueUnterminated(mailbox)
+    .evalTap { case DoConsume =>
+      val process = for {
```

Comment:
maybe extract this to a method? reading this it's hard to follow, so we need to somehow break things down